pyro/infer/autoguide/guides.py

```diff
@@ -583,6 +583,133 @@ def _loc_scale(self, *args, **kwargs):
+        svi = SVI(model, guide, ...)
+        for _ in range(100):
+            svi.step(data)
+
+        # Move to stage 2.
+        guide.init_scale()
+        for _ in range(100):
+            svi.step(data)
+
+        # Move to stage 3.
+        guide.init_cov()
+        for _ in range(100):
+            svi.step(data)
+
+    The guide does not depend on the model's ``*args, **kwargs``.
+
+    :param callable model: A generative model.
+    :param int rank: The rank of the low-rank part of the covariance matrix.
+    :param str prefix: A prefix for all internal param sites.
+    :param callable init_loc_fn: A per-site initialization function.
+        See :ref:`autoguide-initialization` section for available functions.
+    """
+
+    def __init__(self, model, prefix="auto", init_loc_fn=init_to_median, rank=1):
+        if not isinstance(rank, int) or not rank > 0:
+            raise ValueError("Expected rank > 0 but got {}".format(rank))
+        self.rank = rank
+        self._init_scale = 0.
+        self._init_cov = 0.
+        super(AutoProgressive, self).__init__(model, prefix=prefix, init_loc_fn=init_loc_fn)
+
+    def init_scale(self, scale=0.1):
+        """
+        Initialize diagonal uncertainty scale.
+        This is typically called after a point estimate has been learned.
+
+        :param float scale: An initial uncertainty scale. Must be postive.
+            Values less than 1 are recommended.
+        """
+        if not isinstance(scale, float) or not scale > 0.:
+            raise ValueError("Expected positive init_scale but got {}".format(scale))
+        if self._init_scale > 0.:
+            raise ValueError("Cannot initialize scale a second time.")
+        self._init_scale = scale
+
+    def init_cov(self, scale=0.1):
+        """
+        Initialize low rank covariance factor to scaled random noise.
+        This is typically called after both point estimates and noise scales
+        have been learned; the initial low rank ``cov_factor`` scale combines
+        the ``scale`` arg with the current learned diagonal noise scales.
+
+        :param float scale: A noise scale. Must be positive.
+            Values less than 1 are recommended.
+        """
+        if not isinstance(scale, float) or not scale > 0.:
+            raise ValueError("Expected positive init_cov but got {}".format(scale))
+        if self._init_cov > 0.:
+            raise ValueError("Cannot initialize cov_factor a second time.")
+        if self._init_scale == 0.:
+            raise ValueError("Cannot initialize cov_factor before initializing scale.")
+        self._init_cov = scale
+
+    def get_posterior(self, *args, **kwargs):
+        """
+        Returns a posterior distribution over the latent unconstrained value.
+        """
+        loc = pyro.param("{}_loc".format(self.prefix), self._init_loc)
+        if self._init_scale == 0.:
+            return dist.Delta(loc, event_dim=1)  # Stage 1.
+        assert self._init_scale > 0.
```

Comment:
is this missing the MVN return?

Reply:
This only ever returns `Delta`, `Independent(Normal)`, or `LowRankMultivariateNormal` 😕